pkg/destroy/kubevirt/destroyer.go

```diff
@@ -0,0 +1,152 @@
+package kubevirt
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apilabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	ickubevirt "github.com/openshift/installer/pkg/asset/installconfig/kubevirt"
+	"github.com/openshift/installer/pkg/destroy/providers"
+	"github.com/openshift/installer/pkg/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterUninstaller holds the Metadata info needed to delete the tenantCluster resources from the infraCluster.
+type ClusterUninstaller struct {
+	Metadata types.ClusterMetadata
+	Logger   logrus.FieldLogger
+}
+
+// New returns KubeVirt Uninstaller from ClusterMetadata.
+func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.Destroyer, error) {
+	return &ClusterUninstaller{
+		Metadata: *metadata,
+		Logger:   logger,
+	}, nil
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (uninstaller *ClusterUninstaller) Run() error {
+	namespace := uninstaller.Metadata.Kubevirt.Namespace
+
+	listOpts := metav1.ListOptions{LabelSelector: apilabels.FormatLabels(uninstaller.Metadata.Kubevirt.Labels)}
+	kubevirtClient, err := ickubevirt.NewClient()
+	if err != nil {
+		return err
+	}
+
+	if err := uninstaller.deleteAllVMs(namespace, listOpts, kubevirtClient); err != nil {
+		uninstaller.Logger.Errorf("failed to delete VirtualMachines: %s", err)
+		return err
+	}
+	if err := uninstaller.deleteAllDVs(namespace, listOpts, kubevirtClient); err != nil {
+		uninstaller.Logger.Errorf("failed to delete DataVolumes: %s", err)
+		return err
+	}
+	if err := uninstaller.deleteAllSecrets(namespace, listOpts, kubevirtClient); err != nil {
+		uninstaller.Logger.Errorf("failed to delete Secrets: %s", err)
+		return err
+	}
+	return nil
+}
+
+func (uninstaller *ClusterUninstaller) deleteAllVMs(namespace string, listOpts metav1.ListOptions, kubevirtClient ickubevirt.Client) error {
+	vmList, err := kubevirtClient.ListVirtualMachine(namespace, listOpts)
+	if err != nil {
+		return err
+	}
+
+	if len(vmList.Items) == 0 {
+		uninstaller.Logger.Infof("There is no Virtual Machine to delete in namespace %s", namespace)
+		return nil
+	}
+	uninstaller.Logger.Infof("Found %d Virtual Machines to delete in namespace %s", len(vmList.Items), namespace)
+	for _, vm := range vmList.Items {
+		uninstaller.Logger.Infof("Delete Virtual Machine %s from Namespace %s", vm.Name, namespace)
+		if err := kubevirtClient.DeleteVirtualMachine(namespace, vm.Name); err != nil {
+			return err
```

Comment:
The destroyer cannot just abort when there is a problem with a deletion attempt. The destroyer must continue to attempt to destroy indefinitely. Also, if there is a problem destroying one resource, then that should not block the destroying from destroying other resources.

Reply:
Change it - the destroyer collect all the errors and return error only after all resources destroy were called

Reply:
There is still the problem that the destroyer is not retrying deletes that failed.

Reply:
Hi @staebler 
I changed the destroyer to work with channels and in case of errors retry to destroy indefinitely
Please re-review it